Dump all state to JSON

cwgoes opened the pull request:
Closes https://github.com/cosmos/cosmos-sdk/issues/131

cwgoes commented:
Swapped to general `export` command, so usage would be like:

```bash
gaiad export > ~/.gaiad/config/genesis.json
# modify some consensus rules, rebuild
gaiad start
```

cwgoes commented:
Ready for review, the code coverage isn't very accurate & this isn't consensus-level anyways.

Thoughts on this approach / alternative suggestions welcome.

rigelrozanski commented on x/stake/keeper.go:
```diff
@@ -371,6 +371,30 @@ func (k Keeper) GetDelegatorBond(ctx sdk.Context,
 }
 
+// load all bonds
+func (k Keeper) GetBonds(ctx sdk.Context, maxRetrieve int16) (bonds []DelegatorBond) {
```
Does this need to be exposed? I don't think we need to use this outside of specialized stake functionality

cwgoes replied:
Nope, now unexposed.

rigelrozanski commented on x/stake/handler.go:
```diff
@@ -49,9 +49,29 @@ func NewEndBlocker(k Keeper) sdk.EndBlocker {
 
 // InitGenesis - store genesis parameters
-func InitGenesis(ctx sdk.Context, k Keeper, data GenesisState) {
+func (k Keeper) InitGenesis(ctx sdk.Context, data GenesisState) {
```
`InitGenesis` (and `WriteGenesis` now) should not be functions of the keeper. It's subtle, but a while back Jae and I came the conclusion that handler ought to not be a function of keeper, by extension I'd put genesis functionality in the same category

rigelrozanski commented on cmd/gaia/app/app.go:
```diff
@@ -149,7 +149,15 @@ func (app *GaiaApp) initChainer(ctx sdk.Context, req abci.RequestInitChain) abci
 
 	// load the initial stake information
-	stake.InitGenesis(ctx, app.stakeKeeper, genesisState.StakeData)
+	stake.InitGenesis(app.stakeKeeper, ctx, genesisState.StakeData)
```
why did you switch it's standard that the context is always the first function input

cwgoes replied:
OK, updated all instances.

rigelrozanski commented on cmd/gaia/cmd/gaiad/main.go:
```diff
@@ -39,3 +40,16 @@ func generateApp(rootDir string, logger log.Logger) (abci.Application, error) {
+	if err != nil {
+		return nil, nil, err
+	}
+	bapp := app.NewGaiaApp(log.NewNopLogger(), db)
```
bapp -> gapp

cwgoes replied:
Updated.

cwgoes commented on cmd/gaia/cmd/gaiad/main.go:
```diff
@@ -39,3 +40,16 @@ func generateApp(rootDir string, logger log.Logger) (abci.Application, error) {
+	}
+	bapp := app.NewGaiaApp(log.NewNopLogger(), db)
+	if err != nil {
+		return nil, nil, err
```
Oops, too much vim-copy-search-replace fu. Fixed.

cwgoes commented on cmd/gaia/cmd/gaiad/main.go:
```diff
@@ -39,3 +40,16 @@ func generateApp(rootDir string, logger log.Logger) (abci.Application, error) {
 	return bapp, nil
 }
+
+func exportApp(rootDir string, logger log.Logger) (interface{}, *wire.Codec, error) {
```
I agree, I'll do this in this PR.

cwgoes replied:
Reworked `generateApp` and `baseApp` to share common functionality, let me know what you think.

cwgoes commented:
Addressed PR comments and added an iterator to export accounts properly. Testsuite presently fails, I think because the account mapper no longer uses the main store.
